core/templates/pages/diagnostic-test-player-page/diagnostic-test-topic-tracker.model.ts

```diff
@@ -0,0 +1,265 @@
+        lastTopicId = unprocessedAncestorTopicIds.pop() || '';
+
+        if (ancestorTopicIds.indexOf(lastTopicId) === -1) {
+          ancestorTopicIds.push(lastTopicId);
+        }
+
+        unprocessedAncestorTopicIds = unprocessedAncestorTopicIds.concat(
+          this._topicIdToPrerequisiteTopicIds[lastTopicId].filter(
+            (topic) => visitedTopicIdsForCurrentTopic.indexOf(topic) === -1
+          ));
+        visitedTopicIdsForCurrentTopic.push(lastTopicId);
+      }
+      this._topicIdToAncestorTopicIds[topicId] = ancestorTopicIds.sort();
+    }
+  }
+
+  generateTopicIdToSuccessorTopicIds(): void {
+    // The method generates a dict with topic ID as the key and all of its
+    // successor topic IDs as value. The successor's list is generated using
+    // the prerequisite dependency graph.
+    // Example: A -> B -> C, here the topic ID to successor topic IDs dict will
+    // look like: {A: [B, C], B: [C], C: []}.
+    let topicIdToChildTopicId: TopicIdToRelatedTopicIds = {};
+
+    for (let topicId in this._topicIdToPrerequisiteTopicIds) {
+      topicIdToChildTopicId[topicId] = [];
+    }
+
+    for (let topicId in this._topicIdToPrerequisiteTopicIds) {
+      let prereqTopicIds = this._topicIdToPrerequisiteTopicIds[topicId];
+      for (let prereqTopicId of prereqTopicIds) {
+        topicIdToChildTopicId[prereqTopicId].push(topicId);
+      }
+    }
+
+    for (let topicId in topicIdToChildTopicId) {
+      let successorTopicIds: string[] = [];
+      let unprocessedSuccessorTopicIds: string[] = cloneDeep(
+        topicIdToChildTopicId[topicId]);
+
+      let visitedTopicIdsForCurrentTopic: string[] = [];
+      let lastTopicId: string;
+
+      while (unprocessedSuccessorTopicIds.length > 0) {
+        lastTopicId = unprocessedSuccessorTopicIds.pop() || '';
+
+        if (successorTopicIds.indexOf(lastTopicId) === -1) {
+          successorTopicIds.push(lastTopicId);
+        }
+
+        unprocessedSuccessorTopicIds = unprocessedSuccessorTopicIds.concat(
+          topicIdToChildTopicId[lastTopicId].filter(
+            (topicId) => visitedTopicIdsForCurrentTopic.indexOf(topicId) === -1
+          ));
+        visitedTopicIdsForCurrentTopic.push(lastTopicId);
+      }
+      this._topicIdToSuccessorTopicIds[topicId] = successorTopicIds.sort();
+    }
+  }
+
+  selectNextTopicIdToTest(): string {
+    // A dict with topic ID as key and
+    // min(length of ancestors, length of successors) topic IDs as value.
+    let topicIdToLengthOfRelatedTopicIds: {[topicId: string]: number} = {};
+
+    for (let topicId of this._eligibleTopicIds) {
+      let ancestorTopicIds = this._topicIdToAncestorTopicIds[topicId];
+      let successorTopicIds = this._topicIdToSuccessorTopicIds[topicId];
+
+      // Considering a given topic, the assured number of topics that will be
+      // removed from the eligible topic IDs after testing, is the minimum
+      // between its ancestors and successors.
```

Comment:
...is the minimum of the counts of its ancestors and successors.

Reply:
Updated